### PR TITLE
Start to Add WASM Support Again

### PR DESCRIPTION
**Pull request by cptpiepmatz**
# Description

The [nushell/demo](https://github.com/nushell/demo) project successfully demonstrated running Nushell in the browser using WASM. However, the current version of Nushell cannot be easily built for the `wasm32-unknown-unknown` target, the default for `wasm-bindgen`.

This PR introduces initial support for the `wasm32-unknown-unknown` target by disabling OS-dependent features such as filesystem access, IO, and platform/system-specific functionality. This separation is achieved using a new `os` feature in the following crates:

 - `nu-cmd-lang`
 - `nu-command`
 - `nu-engine`
 - `nu-protocol`

The `os` feature includes all functionality that interacts with an operating system. It is enabled by default, but can be disabled using `--no-default-features`. All crates that depend on these core crates now use `--no-default-features` to allow compilation for WASM.

To demonstrate compatibility, the following script builds all crates expected to work with WASM. Direct user interaction, running external commands, working with plugins, and features requiring `openssl` are out of scope for now due to their complexity or reliance on C libraries, which are difficult to compile and link in a WASM environment.

```nushell
[ # compatible crates
	"nu-cmd-base",
	"nu-cmd-extra",
	"nu-cmd-lang",
	"nu-color-config",
	"nu-command",
	"nu-derive-value",
	"nu-engine",
	"nu-glob",
	"nu-json",
	"nu-parser",
	"nu-path",
	"nu-pretty-hex",
	"nu-protocol",
	"nu-std",
	"nu-system",
	"nu-table",
	"nu-term-grid",
	"nu-utils",
	"nuon"
] | each {cargo build -p $in --target wasm32-unknown-unknown --no-default-features}
```

## Caveats
This PR has a few caveats:
1. **`miette` and `terminal-size` Dependency Issue**
   `miette` depends on `terminal-size`, which uses `rustix` when the target is not Windows. However, `rustix` requires `std::os::unix`, which is unavailable in WASM. To address this, I opened a [PR](https://github.com/eminence/terminal-size/pull/68) for `terminal-size` to conditionally compile `rustix` only when the target is Unix. For now, the `Cargo.toml` includes patches to:
    - Use my forked version of `terminal-size`.
    - ~~Use an unreleased version of `miette` that depends on `terminal-size@0.4`.~~

   These patches are temporary and can be removed once the upstream changes are merged and released.

2. **Test Output Adjustments**
   Due to the slight bump in the `miette` version, one test required adjustments to accommodate minor formatting changes in the error output, such as shifted newlines.

# User-Facing Changes

This shouldn't break anything but allows using some crates for targeting `wasm32-unknown-unknown` to revive the demo page eventually.

# Tests + Formatting

- :green_circle: `toolkit fmt`
- :green_circle: `toolkit clippy`
- :green_circle: `toolkit test`
- :green_circle: `toolkit test stdlib`

I did not add any extra tests, I just checked that compiling works, also when using the host target but unselecting the `os` feature.

# After Submitting

~~Breaking the wasm support can be easily done by adding some `use`s or by adding a new dependency, we should definitely add some CI that also at least builds against wasm to make sure that building for it keep working.~~
I added a job to build wasm.

### CODE REVIEW

**Comment by fdncred** [comment]
Would it help to remove terminal-size altogether and replace it with crossterm's terminal size? https://github.com/crossterm-rs/crossterm/blob/fc8f9779382b7faa7f92bc40d72f8dcbd5849296/src/terminal.rs#L136

**Comment by cptpiepmatz** [comment]
> Would it help to remove terminal-size altogether and replace it with crossterm's terminal size? https://github.com/crossterm-rs/crossterm/blob/fc8f9779382b7faa7f92bc40d72f8dcbd5849296/src/terminal.rs#L136

`terminal-size` is a depency of `miette`, so we have to deal with it somehow.

**Comment by cptpiepmatz** [comment]
@fdncred I added to the toolkit a new command called `toolkit build wasm` that builds all crates that should work for wasm. Also I added a CI step called `build-wasm` that does the same.

**Comment by fdncred** [comment]
Looks good @cptpiepmatz. Need to get some other @nushell/core-team approvals to bring this back.

**Comment by IanManske** [comment]
Nice work! Would be great to have this working again. I haven't looked over the code yet, but `cargo hack --all --feature-powerset --skip default-no-clipboard,stable,mimalloc check` is failing.

**Comment by cptpiepmatz** [comment]
@IanManske I've got `cargo hack --all --feature-powerset --skip default-no-clipboard,stable,mimalloc check` working, mostly it needed that the `plugin` features include `os`.

**Comment by stormasm** [comment]
Wow this is great !
Especially this concept...

```rust
source toolkit.nu
build wasm
```

Its all working, very sweet 😄

**Comment by fdncred** [comment]
> @IanManske I've got `cargo hack --all --feature-powerset --skip default-no-clipboard,stable,mimalloc check` working, mostly it needed that the `plugin` features include `os`.

It's failing for me on MacOS.
```
info: running `cargo check --no-default-features` on nu-command (119/207)
...
error[E0425]: cannot find value `ULimit` in this scope
   --> crates/nu-command/src/default_context.rs:265:25
    |
265 |         bind_command! { ULimit };
    |                         ^^^^^^ not found in this scope

For more information about this error, try `rustc --explain E0425`.
error: could not compile `nu-command` (lib) due to 1 previous error
error: process didn't exit successfully: `/Users/fdncred/.rustup/toolchains/1.80.1-aarch64-apple-darwin/bin/cargo check --manifest-path crates/nu-command/Cargo.toml --no-default-features` (exit status: 101)
```

**Comment by cptpiepmatz** [comment]
@fdncred I fixed that `ULimit` was disabled on my Windows as the `#[cfg(unix)]` flag did not include it.

**Comment by fdncred** [comment]
I'm testing again.

File: Cargo.toml [hunk 1]
```diff
@@ -317,6 +318,12 @@ bench = false
 # reedline = { git = "https://github.com/nushell/reedline", branch = "main" }
 # nu-ansi-term = {git = "https://github.com/nushell/nu-ansi-term.git", branch = "main"}
 
+# this revision bumped terminal_size to use 0.4.0
+miette = { git = "https://github.com/zkat/miette.git", rev = "21e9a70ad8aacd1f32f515fa1540adc1c8df5341" }
+
+# removed rustix from non-unix builds
+terminal_size = { git = "https://github.com/cptpiepmatz/terminal-size.git" }
```
Comment (IanManske) [inline Cargo.toml:325]: Once this is ready to merge, this should ideally be merged back into `terminal-size` upstream (I saw your PR). Alternatively, this should have a `rev =` specification.

File: crates/nu-command/Cargo.toml [hunk 2]
```diff
@@ -134,7 +135,53 @@ features = [
...
+	"uu_whoami",
+	"which",
+]
+
+# The dependencies listed below need 'getrandom'. 
+# They work with JS (usually with wasm-bindgen) or regular OS support. 
+# Hence they are also put under the 'os' feature to avoid repetition.
+js = [
+	"getrandom", 
+	"getrandom/js",
```
Comment (IanManske) [inline crates/nu-command/Cargo.toml:166]: Is `getrandom/js` necesssary for `os`?

**Review by IanManske — COMMENTED** [review]
Nice work! Relatively straight forward for the most part, just a few commentsL

File: crates/nu-command/Cargo.toml (hunk 2 above)
Comment (cptpiepmatz) [reply crates/nu-command/Cargo.toml:166]: It's not but `os` should get `rand` which is not in `network`. Also `getrandom/js` is a noop for systems that have an OS. See [the docs of `getrandom`](https://docs.rs/getrandom/latest/getrandom/#webassembly-support), about the `js` feature:
> This feature has no effect on targets other than `wasm32-unknown-unknown`.

File: Cargo.toml (hunk 1 above)
Comment (cptpiepmatz) [reply Cargo.toml:325]: I hope they take a look at my PR relatively soon.

Comment (fdncred) [reply Cargo.toml:325]: I believe miette and terminal_size will need to be published before we can release, right? I wonder if it's safe to land this without agreements from them to publish before Dec-24th.

**Comment by cptpiepmatz** [comment]
With `miette@7.3` merged, we don't need that patch for `miette` anymore. Sadly the author of `terminal-size` doesn't seem to be very active.

File: Cargo.toml (hunk 1 above)
Comment (cptpiepmatz) [reply Cargo.toml:322]: `miette` got updated on `main`, we only need the patch to `terminal-size` to go through, but the author doesn't seem to be very active

**Comment by fdncred** [comment]
We could replace terminal-size with something else if we needed to. I haven't looked too deep into other crates.

**Comment by cptpiepmatz** [comment]
> We could replace terminal-size with something else if we needed to. I haven't looked too deep into other crates.

`terminal-size` is a dependency of `miette`, if we want to change it out, we would have to change it for `miette`

**Comment by fdncred** [comment]
double ugh! i forgot and that's the second or third time you've told me that. LOL

File: Cargo.toml [hunk 3]
```diff
@@ -134,6 +134,7 @@ quickcheck = "1.0"
 quickcheck_macros = "1.0"
 quote = "1.0"
 rand = "0.8"
+getrandom = "*" # use version requested by rand
```
Comment (sholderbach) [inline Cargo.toml:137]: I think we can't publish with a star-dependency as well.

**Review by sholderbach — COMMENTED** [review]
Wuhu all green!

File: Cargo.toml (hunk 3 above)
Comment (cptpiepmatz) [reply Cargo.toml:137]: Should I update it to the version that `rand` requests manually and add a comment about it?

File: crates/nu-cmd-extra/Cargo.toml [hunk 4]
```diff
@@ -17,12 +17,12 @@ workspace = true
 
 [dependencies]
 nu-cmd-base = { path = "../nu-cmd-base", version = "0.100.1" }
-nu-engine = { path = "../nu-engine", version = "0.100.1" }
+nu-engine = { path = "../nu-engine", version = "0.100.1", default-features = false }
```
[the remark this answers is not in the document]
Comment (cptpiepmatz) [reply crates/nu-cmd-extra/Cargo.toml:20]: I don't understand. `nu-engine` has a default feature that I want to disable by default.

File: Cargo.toml (hunk 3 above)
Comment (sholderbach) [reply Cargo.toml:137]: Yeah, I think this either needs to be a compatible range or minimal version.

File: crates/nu-cmd-extra/Cargo.toml (hunk 4 above)
[the remark this answers is not in the document]
Comment (sholderbach) [reply crates/nu-cmd-extra/Cargo.toml:20]: I was wondering if all those extra feature annotations could just be part of the main `Cargo.toml` but if I recall that doesn't quite work cleanly.

This diff seems fine to me.

File: Cargo.toml (hunk 3 above)
Comment (cptpiepmatz) [reply Cargo.toml]: `rand` depends on `rand_core` and that depends on `getrandom@0.2`, so I set that in the `Cargo.toml` now

File: crates/nu-cmd-extra/Cargo.toml (hunk 4 above)
[the remark this answers is not in the document]
Comment (cptpiepmatz) [reply crates/nu-cmd-extra/Cargo.toml:20]: I don't see any `nu-*` crates in the `workspace.dependencies` section, so I guess this is the correct solution. Nushell (the binary) wants the default features but other (wasm) users don't want them

[the remark this answers is not in the document]
Comment (sholderbach) [reply crates/nu-cmd-extra/Cargo.toml:20]: Yeah, was just shooting in the dark :D

File: Cargo.toml (hunk 3 above)
Comment (sholderbach) [reply Cargo.toml]: Thanks!

**Comment by fdncred** [comment]
Nice job @cptpiepmatz. We appreciate your hard work here. Hopefully we can get the demo site updated with a new version of nushell.

**Comment by fdncred** [comment]
@cptpiepmatz I'm not getting the plugins listed when I run the `version` command. There is no `installed_plugins` key. I think it's due to this PR. Would you mind investigating it. I noticed it on Windows, but it may be in other platforms too.

**Comment by cptpiepmatz** [comment]
> @cptpiepmatz I'm not getting the plugins listed when I run the `version` command. There is no `installed_plugins` key. I think it's due to this PR. Would you mind investigating it. I noticed it on Windows, but it may be in other platforms too.

@fdncred I have a feeling where the issue comes from, taking a look right now.

**Pull request merged**